cluster/manager.go

```diff
@@ -0,0 +1,330 @@
+	totalPrimaryNotReady = stats.NewGauge32("cluster.total.state.primary-not-ready")
+	// metric cluster.total.state.secondary-ready is the number of nodes we know to be secondary and ready
+	totalSecondaryReady = stats.NewGauge32("cluster.total.state.secondary-ready")
+	// metric cluster.total.state.secondary-not-ready is the number of nodes we know to be secondary and not ready
+	totalSecondaryNotReady = stats.NewGauge32("cluster.total.state.secondary-not-ready")
+	// metric cluster.total.partitions is the number of partitions in the cluster that we know of
+	totalPartitions = stats.NewGauge32("cluster.total.partitions")
+
+	// metric cluster.decode_err.join is a counter of json unmarshal errors
+	unmarshalErrJoin = stats.NewCounter32("cluster.decode_err.join")
+	// metric cluster.decode_err.update is a counter of json unmarshal errors
+	unmarshalErrUpdate = stats.NewCounter32("cluster.decode_err.update")
+	// metric cluster.decode_err.merge_remote_state is a counter of json unmarshal errors
+	unmarshalErrMergeRemoteState = stats.NewCounter32("cluster.decode_err.merge_remote_state")
+)
+
+type ClusterManager struct {
+	sync.RWMutex
+	Peers map[string]Node
+	node  Node
+	list  *memberlist.Memberlist
+}
+
+func (c *ClusterManager) setList(list *memberlist.Memberlist) {
+	c.Lock()
+	c.list = list
+	c.Unlock()
+}
+
+func (c *ClusterManager) ThisNode() Node {
+	c.RLock()
+	defer c.RUnlock()
+	return c.node
+}
+
+func (c *ClusterManager) PeersList() []Node {
+	c.RLock()
+	list := make([]Node, len(c.Peers), len(c.Peers))
+	i := 0
+	for _, p := range c.Peers {
+		list[i] = p
+		i++
+	}
+	c.RUnlock()
+	return list
+}
+
+// report the cluster stats every time there is a change to the cluster state.
+// it is assumed that the lock is acquired before calling this method.
+func (c *ClusterManager) clusterStats() {
+	primReady := 0
+	primNotReady := 0
+	secReady := 0
+	secNotReady := 0
+	partitions := make(map[int32]int)
+	for _, p := range c.Peers {
+		if p.Primary {
+			if p.IsReady() {
+				primReady++
+			} else {
+				primNotReady++
+			}
+		} else {
+			if p.IsReady() {
+				secReady++
+			} else {
+				secNotReady++
+			}
+		}
+		for _, partition := range p.Partitions {
+			partitions[partition]++
+		}
+	}
+
+	totalPrimaryReady.Set(primReady)
+	totalPrimaryNotReady.Set(primNotReady)
+	totalSecondaryReady.Set(secReady)
+	totalSecondaryNotReady.Set(secNotReady)
+
+	totalPartitions.Set(len(partitions))
+}
+
+func (c *ClusterManager) NotifyJoin(node *memberlist.Node) {
+	eventsJoin.Inc()
+	c.Lock()
+	defer c.Unlock()
+	if len(node.Meta) == 0 {
+		return
+	}
+	log.Info("CLU manager: Node %s with address %s has joined the cluster", node.Name, node.Addr.String())
+	peer := Node{}
+	err := json.Unmarshal(node.Meta, &peer)
+	if err != nil {
+		log.Error(3, "CLU manager: Failed to decode node meta from %s: %s", node.Name, err.Error())
+		unmarshalErrJoin.Inc()
+		return
+	}
+	peer.RemoteAddr = node.Addr.String()
+	if peer.Name == c.node.Name {
```

Comment:
if you accidentally start an instance with the same name, it'll be marked as local despite having a remote remoteaddr. is that a good idea? (same for notifyUpdate)